File: apps/android/app/src/test/java/com/screenmesh/crypto/IdentityTest.kt
package com.screenmesh.crypto

import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/** Kotlin mirror of packages/crypto/src/identity.test.ts. */
class IdentityTest {
    @Test
    fun `generateIdentity gives every identity a distinct deviceId and key material`() {
        val a = generateIdentity()
        val b = generateIdentity()
        assertNotEquals(a.deviceId, b.deviceId)
        assertNotEquals(exportPublicKey(a.publicKey), exportPublicKey(b.publicKey))
    }

    @Test
    fun `verify accepts a signature from the matching identity`() {
        val identity = generateIdentity()
        val data = "hello screenmesh".toByteArray()
        val signature = sign(identity, data)
        assertTrue(verify(identity.publicKey, signature, data))
    }

    @Test
    fun `verify rejects a signature checked against a different device's public key`() {
        val identity = generateIdentity()
        val impostor = generateIdentity()
        val data = "hello screenmesh".toByteArray()
        val signature = sign(identity, data)
        assertFalse(verify(impostor.publicKey, signature, data))
    }

    @Test
    fun `verify rejects a signature when the signed data has been tampered with`() {
        val identity = generateIdentity()
        val signature = sign(identity, "original".toByteArray())
        assertFalse(verify(identity.publicKey, signature, "tampered".toByteArray()))
    }

    @Test
    fun `exportPublicKey and importPublicKey round-trip and still verify`() {
        val identity = generateIdentity()
        val reimported = importPublicKey(exportPublicKey(identity.publicKey))
        val data = "round trip check".toByteArray()
        val signature = sign(identity, data)
        assertTrue(verify(reimported, signature, data))
    }

    @Test
    fun `exportEd25519PrivateKey and importEd25519PrivateKey round-trip and can still sign`() {
        // Mirrors on-disk identity persistence — a bug here would silently
        // corrupt identity across an app relaunch (see sync/LocalState.kt).
        val identity = generateIdentity()
        val reimported = importEd25519PrivateKey(exportEd25519PrivateKey(identity.privateKey))
        val data = "persisted identity check".toByteArray()
        val signature = sign(identity.copy(privateKey = reimported), data)
        assertTrue(verify(identity.publicKey, signature, data))
    }

    @Test
    fun `exportEncryptionPublicKey and importEncryptionPrivateKey round-trip for X25519 agreement`() {
        val alice = generateIdentity()
        val bob = generateIdentity()

        val bobPubReimported = importEncryptionPublicKey(exportEncryptionPublicKey(bob.encryptionPublicKey))
        val alicePrivReimported =
            importEncryptionPrivateKey(exportEncryptionPrivateKey(alice.encryptionPrivateKey))

        // Two independent ECDH computations from reimported keys must agree —
        // this is exactly what the ratchet bootstrap depends on.
        val agreementA = run {
            val agreement = org.bouncycastle.crypto.agreement.X25519Agreement()
            agreement.init(alicePrivReimported)
            val out = ByteArray(agreement.agreementSize)
            agreement.calculateAgreement(bobPubReimported, out, 0)
            out
        }
        val agreementB = run {
            val agreement = org.bouncycastle.crypto.agreement.X25519Agreement()
            agreement.init(bob.encryptionPrivateKey)
            val out = ByteArray(agreement.agreementSize)
            agreement.calculateAgreement(alice.encryptionPublicKey, out, 0)
            out
        }
        org.junit.Assert.assertArrayEquals(agreementA, agreementB)
    }
}

File: packages/crypto/src/identity.test.ts
import { describe, expect, it } from "vitest";
import {
  exportEd25519PrivateKey,
  exportPublicKey,
  generateIdentity,
  importEd25519PrivateKey,
  importPublicKey,
  sign,
  verify,
} from "./identity.js";

describe("generateIdentity", () => {
  it("gives every identity a distinct deviceId and distinct key material", async () => {
    const a = await generateIdentity();
    const b = await generateIdentity();
    expect(a.deviceId).not.toBe(b.deviceId);
    expect(await exportPublicKey(a.publicKey)).not.toBe(await exportPublicKey(b.publicKey));
  });
});

describe("sign / verify", () => {
  it("verifies a signature produced by the matching identity", async () => {
    const identity = await generateIdentity();
    const data = new TextEncoder().encode("hello screenmesh");
    const signature = await sign(identity, data);
    expect(await verify(identity.publicKey, signature, data)).toBe(true);
  });

  it("rejects a signature checked against a different device's public key", async () => {
    const identity = await generateIdentity();
    const impostor = await generateIdentity();
    const data = new TextEncoder().encode("hello screenmesh");
    const signature = await sign(identity, data);
    expect(await verify(impostor.publicKey, signature, data)).toBe(false);
  });

  it("rejects a signature when the signed data has been tampered with", async () => {
    const identity = await generateIdentity();
    const signature = await sign(identity, new TextEncoder().encode("original"));
    expect(await verify(identity.publicKey, signature, new TextEncoder().encode("tampered"))).toBe(
      false,
    );
  });
});

describe("exportPublicKey / importPublicKey", () => {
  it("round-trips a public key through base64 and still verifies signatures", async () => {
    const identity = await generateIdentity();
    const reimported = await importPublicKey(await exportPublicKey(identity.publicKey));
    const data = new TextEncoder().encode("round trip check");
    const signature = await sign(identity, data);
    expect(await verify(reimported, signature, data)).toBe(true);
  });
});

describe("exportEd25519PrivateKey / importEd25519PrivateKey", () => {
  it("round-trips an extractable private key and can still sign with it", async () => {
    // Mirrors the desktop agent's on-disk persistence path (no IndexedDB
    // to lean on) — a bug here would silently corrupt identity on reload.
    const identity = await generateIdentity({ extractable: true });
    const exported = await exportEd25519PrivateKey(identity.privateKey);
    const reimported = await importEd25519PrivateKey(exported);

    const data = new TextEncoder().encode("persisted identity check");
    const signature = await sign({ ...identity, privateKey: reimported }, data);
    expect(await verify(identity.publicKey, signature, data)).toBe(true);
  });
});
